pkg/genericcli/context.go

```diff
@@ -0,0 +1,840 @@
+	}
+
+	// fallback to the default specified by viper
+	return viper.GetString(keyAPIURL)
+}
+
+func (c *Context) GetProvider() string {
+	if viper.IsSet(keyProvider) {
+		return viper.GetString(keyProvider)
+	}
+	return c.Provider
+}
+
+func (cs *contexts) validate() error {
+	names := map[string]bool{}
+	for _, context := range cs.Contexts {
+		names[context.Name] = true
+		if context.Name == "" {
+			return fmt.Errorf(errMsgBlankContextField, "Name")
+		}
+		if context.APIToken == "" {
+			return fmt.Errorf(errMsgBlankContextField, "APIToken")
+		}
+	}
+
+	if len(cs.Contexts) != len(names) {
+		return errContextNamesAreUnique
+	}
+
+	return nil
+}
+
+func (cs *contexts) delete(name string) *Context {
+	var deletedCtx *Context
+	cs.Contexts = slices.DeleteFunc(cs.Contexts, func(ctx *Context) bool {
+		if ctx.Name == name {
+			deletedCtx = ctx
+		}
+		return ctx.Name == name
+	})
+
+	return deletedCtx
+}
+
+func (cs *contexts) getByName(name string) (*Context, bool) {
+	for _, context := range cs.Contexts {
+		if context.Name == name {
+			return context, true
+		}
+	}
+	return nil, false
+}
+
+func contextSorter() *multisort.Sorter[*Context] {
+	return multisort.New(multisort.FieldMap[*Context]{
+		sortKeyName: func(a, b *Context, descending bool) multisort.CompareResult {
+			return multisort.Compare(a.Name, b.Name, descending)
+		},
+		sortKeyAPIURL: func(a, b *Context, descending bool) multisort.CompareResult {
+			urlA := pointer.SafeDeref(a.APIURL)
+			urlB := pointer.SafeDeref(b.APIURL)
+			return multisort.Compare(urlA, urlB, descending)
+		},
+		sortKeyDefaultProject: func(a, b *Context, descending bool) multisort.CompareResult {
+			return multisort.Compare(a.DefaultProject, b.DefaultProject, descending)
+		},
+		sortKeyTimeout: func(a, b *Context, descending bool) multisort.CompareResult {
+			timeoutA := pointer.SafeDeref(a.Timeout)
+			timeoutB := pointer.SafeDeref(b.Timeout)
+			return multisort.Compare(timeoutA, timeoutB, descending)
+		},
+		sortKeyProvider: func(a, b *Context, descending bool) multisort.CompareResult {
+			return multisort.Compare(a.Provider, b.Provider, descending)
+		},
+	}, multisort.Keys{{ID: sortKeyName}})
+}
+
+func (c *ContextManager) GetContextCurrentOrDefault() *Context {
+	ctxs, err := c.getContexts()
+	if err != nil {
+		return defaultCtx()
+	}
+	ctx, ok := ctxs.getByName(ctxs.CurrentContext)
+	if !ok {
+		return defaultCtx()
+	}
+	// TODO deep copy?
+	// return ctx.deepCopy()
+	return ctx
+}
+
+func defaultCtx() *Context {
+	return &Context{
+		Name:     DefaultContextName,
+		APIURL:   pointer.PointerOrNil(viper.GetString(keyAPIURL)),
+		APIToken: viper.GetString(keyAPIToken),
+	}
+}
+
+func (c *ContextManager) GetCurrentContext() (*Context, error) {
+	ctxList, err := c.List()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ctx := range ctxList {
+		if ctx.IsCurrent {
+			return ctx, nil
+		}
+	}
+	return nil, nil
+}
+
+func DefaultContext(c *ContextManager) (*Context, error) {
```

Comment:
This function irritates me:

- what is it used for?
- is it really required in the lib?
- it seems to get the current context in multiple ways
- the naming
- why isn't this is method of `*ContextManager`?